MongoDB adaptor panics when CTRL+C'd

## Bug report

### System info:

- Transporter version
https://github.com/compose/transporter/commit/d2818a5128c7fb7247828ea7069de86ee3cf5fd0
- OS
OS X
- DB version(s)
MongoDB 3.2

### Reproducible Steps:

1. `transporter run`
2. Ctrl+C while running

### What did you expect to happened?
transporter to cleanly shut down

### What actually happened?
the mongodb adaptor panic'd

```
panic: send on closed channel

goroutine 52 [running]:
github.com/compose/transporter/adaptor/mongodb.(*Reader).iterateCollection.func1(0xc42026f040, 0xc4201bbb60, 0xc4201bba40, 0xc4201e8000, 0xc4201bb9e0, 0xc4201ba600)
	/Users/JP/gocode/src/github.com/compose/transporter/adaptor/mongodb/reader.go:152 +0x5a6
created by github.com/compose/transporter/adaptor/mongodb.(*Reader).iterateCollection
	/Users/JP/gocode/src/github.com/compose/transporter/adaptor/mongodb/reader.go:175 +0x93
```

It looks like after the changes to have each collection copy in its own goroutine, a close on the main `out` channel does not cause those goroutines to shutdown cleanly.

--- a/adaptor/mongodb/adaptor_test.go
+++ b/adaptor/mongodb/adaptor_test.go
@@ -20,7 +20,7 @@ var (
 
 	dbsToTest = []*TestData{
 		bulkTestData,
-		readerTestData, filteredReaderTestData,
+		readerTestData, filteredReaderTestData, cancelledReaderTestData,
 		writerTestData,
 		tailTestData}
 )

--- a/adaptor/mongodb/reader.go
+++ b/adaptor/mongodb/reader.go
@@ -1,6 +1,7 @@
 package mongodb
 
 import (
+	"errors"
 	"fmt"
 	"strings"
 	"sync"
@@ -62,32 +63,27 @@ func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
 				log.With("db", r.db).Errorf("unable to list collections, %s", err)
 				return
 			}
-			iterationComplete := r.iterateCollection(session.Copy(), collections, out, done)
 			var wg sync.WaitGroup
-			func() {
-				for {
-					select {
-					case <-done:
-						return
-					case i, ok := <-iterationComplete:
-						if !ok {
+			for _, c := range collections {
+				oplogTime := timeAsMongoTimestamp(time.Now())
+				if err := r.iterateCollection(session.Copy(), c, out, done); err != nil {
+					log.With("db", r.db).Errorln(err)
+					return
+				}
+				log.With("db", r.db).With("collection", c).Infoln("iterating complete")
+				if r.tail {
+					wg.Add(1)
+					log.With("collection", c).Infof("oplog start timestamp: %d", oplogTime)
+					go func(wg *sync.WaitGroup, c string, o bson.MongoTimestamp) {
+						defer wg.Done()
+						errc := r.tailCollection(c, session.Copy(), o, out, done)
+						for err := range errc {
+							log.With("db", r.db).With("collection", c).Errorln(err)
 							return
 						}
-						log.With("db", r.db).With("collection", i.c).Infoln("iterating complete")
-						if r.tail {
-							wg.Add(1)
-							go func(wg *sync.WaitGroup, c string, o bson.MongoTimestamp) {
-								defer wg.Done()
-								errc := r.tailCollection(c, session.Copy(), o, out, done)
-								for err := range errc {
-									log.With("db", r.db).With("collection", c).Errorln(err)
-									return
-								}
-							}(&wg, i.c, i.oplogTime)
-						}
-					}
+					}(&wg, c, oplogTime)
 				}
-			}()
+			}
 			log.With("db", r.db).Infoln("Read completed")
 			// this will block if we're tailing
 			wg.Wait()
@@ -98,82 +94,78 @@ func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
 	}
 }
 
-func (r *Reader) listCollections(mgoSession *mgo.Session, filterFn func(name string) bool) (<-chan string, error) {
-	out := make(chan string)
+func (r *Reader) listCollections(mgoSession *mgo.Session, filterFn func(name string) bool) ([]string, error) {
+	defer mgoSession.Close()
+	var colls []string
 	collections, err := mgoSession.DB(r.db).CollectionNames()
 	if err != nil {
-		return out, err
+		return colls, err
 	}
 	log.With("db", r.db).With("num_collections", len(collections)).Infoln("collection count")
-	go func() {
-		defer func() {
-			mgoSession.Close()
-			close(out)
-		}()
-		for _, c := range collections {
-			if filterFn(c) && !strings.HasPrefix(c, "system.") {
-				log.With("db", r.db).With("collection", c).Infoln("sending for iteration...")
-				out <- c
-			} else {
-				log.With("db", r.db).With("collection", c).Infoln("skipping iteration...")
-			}
+	for _, c := range collections {
+		if filterFn(c) && !strings.HasPrefix(c, "system.") {
+			log.With("db", r.db).With("collection", c).Infoln("adding for iteration...")
+			colls = append(colls, c)
+		} else {
+			log.With("db", r.db).With("collection", c).Infoln("skipping iteration...")
 		}
-		log.With("db", r.db).Infoln("done iterating collections")
-	}()
-	return out, nil
+	}
+	log.With("db", r.db).Infoln("done iterating collections")
+	return colls, nil
 }
 
-func (r *Reader) iterateCollection(mgoSession *mgo.Session, in <-chan string, out chan<- message.Msg, done chan struct{}) <-chan iterationComplete {
-	collectionDone := make(chan iterationComplete)
+func (r *Reader) iterateCollection(s *mgo.Session, c string, out chan<- message.Msg, done chan struct{}) error {
+	iter := r.iterate(s, c)
+	for {
+		select {
+		case msg, ok := <-iter:
+			if !ok {
+				return nil
+			}
+			out <- msg
+		case <-done:
+			return errors.New("iteration cancelled")
+		}
+	}
+}
+
+func (r *Reader) iterate(s *mgo.Session, c string) <-chan message.Msg {
+	msgChan := make(chan message.Msg)
 	go func() {
 		defer func() {
-			mgoSession.Close()
-			close(collectionDone)
+			s.Close()
+			close(msgChan)
 		}()
+		canReissueQuery := r.requeryable(c, s)
+		var lastID interface{}
 		for {
-			select {
-			case c, ok := <-in:
-				if !ok {
-					return
+			log.With("collection", c).Infoln("iterating...")
+			session := s.Copy()
+			iter := r.catQuery(c, lastID, session).Iter()
+			var result bson.M
+			for iter.Next(&result) {
+				if id, ok := result["_id"]; ok {
+					lastID = id
 				}
-				log.With("collection", c).Infoln("iterating...")
-				canReissueQuery := r.requeryable(c, mgoSession)
-				var lastID interface{}
-				oplogTime := timeAsMongoTimestamp(time.Now())
-				log.With("collection", c).Infof("setting oplog start timestamp: %d", oplogTime)
-				for {
-					s := mgoSession.Copy()
-					iter := r.catQuery(c, lastID, s).Iter()
-					var result bson.M
-					for iter.Next(&result) {
-						if id, ok := result["_id"]; ok {
-							lastID = id
-						}
-						out <- message.From(ops.Insert, c, data.Data(result))
-						result = bson.M{}
-					}
-					if err := iter.Err(); err != nil {
-						log.With("database", r.db).With("collection", c).Errorf("error reading, %s", err)
-						s.Close()
-						if canReissueQuery {
-							log.With("database", r.db).With("collection", c).Errorln("attempting to reissue query")
-							time.Sleep(5 * time.Second)
-							continue
-						}
-						break
-					}
-					iter.Close()
-					s.Close()
-					break
+				msgChan <- message.From(ops.Insert, c, data.Data(result))
+				result = bson.M{}
+			}
+			if err := iter.Err(); err != nil {
+				log.With("database", r.db).With("collection", c).Errorf("error reading, %s", err)
+				session.Close()
+				if canReissueQuery {
+					log.With("database", r.db).With("collection", c).Errorln("attempting to reissue query")
+					time.Sleep(5 * time.Second)
+					continue
 				}
-				collectionDone <- iterationComplete{oplogTime, c}
-			case <-done:
-				log.With("db", r.db).Infoln("iterating no more")
 				return
 			}
+			iter.Close()
+			session.Close()
+			return
 		}
 	}()
-	return collectionDone
+	return msgChan
 }
 
 func (r *Reader) catQuery(c string, lastID interface{}, mgoSession *mgo.Session) *mgo.Query {
@@ -198,7 +190,7 @@ func (r *Reader) requeryable(c string, mgoSession *mgo.Session) bool {
 			var result bson.M
 			err := mgoSession.DB(r.db).C(c).Find(nil).Select(bson.M{"_id": 1}).One(&result)
 			if err != nil {
-				fmt.Printf("[ERROR] unable to sample document, %s", err)
+				log.With("database", r.db).With("collection", c).Errorf("unable to sample document, %s", err)
 				break
 			}
 			if id, ok := result["_id"]; ok && sortable(id) {

--- a/adaptor/mongodb/reader_test.go
+++ b/adaptor/mongodb/reader_test.go
@@ -15,8 +15,9 @@ import (
 )
 
 var (
-	readerTestData         = &TestData{"reader_test", "foo", 10}
-	filteredReaderTestData = &TestData{"filtered_reader_test", "foo", 10}
+	readerTestData          = &TestData{"reader_test", "foo", 10}
+	filteredReaderTestData  = &TestData{"filtered_reader_test", "foo", 10}
+	cancelledReaderTestData = &TestData{"cancelled_reader_test", "foo", 100}
 )
 
 var filterFunc = func(c string) bool {
@@ -49,6 +50,10 @@ func TestRead(t *testing.T) {
 }
 
 func TestFilteredRead(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping TestFilteredRead in short mode")
+	}
+
 	reader := newReader(
 		filteredReaderTestData.DB,
 		false,
@@ -76,7 +81,37 @@ func TestFilteredRead(t *testing.T) {
 	close(done)
 }
 
+func TestCancelledRead(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping TestCancelledRead in short mode")
+	}
+
+	reader := newReader(cancelledReaderTestData.DB, false, DefaultCollectionFilter)
+	readFunc := reader.Read(filterFunc)
+	done := make(chan struct{})
+	msgChan, err := readFunc(defaultSession, done)
+	if err != nil {
+		t.Fatalf("unexpected Read error, %s\n", err)
+	}
+	go func() {
+		time.Sleep(1 * time.Second)
+		close(done)
+	}()
+	var numMsgs int
+	for _ = range msgChan {
+		time.Sleep(100 * time.Millisecond)
+		numMsgs++
+	}
+	if numMsgs == cancelledReaderTestData.InsertCount {
+		t.Errorf("bad message count, expected less than %d but got that", cancelledReaderTestData.InsertCount)
+	}
+}
+
 func TestReadRestart(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping TestReadRestart in short mode")
+	}
+
 	var db = "restart_read_test"
 
 	c := &Client{

--- a/pipe/pipe.go
+++ b/pipe/pipe.go
@@ -111,7 +111,6 @@ func (m *Pipe) Stop() {
 
 		// we only worry about the stop channel if we're in a listening loop
 		if m.listening {
-			close(m.In)
 			if len(m.Out) > 0 {
 				close(m.Err)
 			}

--- a/pipeline/pipeline_integration_test.go
+++ b/pipeline/pipeline_integration_test.go
@@ -60,7 +60,7 @@ func TestFileToFile(t *testing.T) {
 	p.Stop()
 	time.Sleep(1 * time.Second)
 	numgorosAfter := runtime.NumGoroutine()
-	if numgorosBefore != numgorosAfter {
+	if numgorosBefore < numgorosAfter {
 		trace := make([]byte, 10240)
 		runtime.Stack(trace, true)
 		t.Errorf("leaky goroutines detected, started with %d, ended with %d\n%s", numgorosBefore, numgorosAfter, trace)

